test/test_nn.py

```diff
@@ -4220,6 +4220,9 @@ def fn(input):
                     out1 = wrapped_m(input)
                     return out0 + out1
 
+                # Make sure we can compute gradients wrt to all the parameters in the case
+                # of double forward
+                fn(input.clone().requires_grad_()).sum().backward()
```

Comment:
It would be better indeed, but its not possible unless we have a functional version of the nn module

Reply:
Gradcheck is amazing ;) 
You can give it inputs and then use them as captures and it works!
We actually do gradcheck our Modules.
You can use a lambda that takes the parameters as input and just drop them. And give that lambda to gradcheck

Reply:
Haha I know :P
but I did try something like that...
The issue is I don't think gradcheck modifies the inputs in-place (which wouldn't work anyway for leaf tensors), and instead it produce the perturbed input out-of-place and then call the function again? I recall the numerical Jacobian becoming zero due to this reason
Alternatively I also tried replacing the entire parameter in the lambda every time with the new value, but this also means that 1) we'd need to compute analytical before numerical Jacobian 2) analytical still wouldn't work because the original input we consider the input is no longer part of the graph in the new module forward computation

But maybe I'm missing something - is there a link to where we test our modules with gradcheck?

Reply:
hmm wait actually it is done in-place

Reply:
Seems like the reason gradcheck fails is that in training mode (if we do power iteration), power iteration updates _u based on a function of the current weight, which is perturbed by a different one-hot vector every time for finite differencing. This is why having the wrapper function set `_u` to a saved clone does not work. 

Due to this interaction, it is probably not possible to test the double forward in-place issue with gradcheck. (Using eval mode to get gradcheck to work here would skip the in-place update to _u entirely which is what we wanted to check in the first place.)